db/snapshot/backup_task.cc

```diff
@@ -48,49 +48,69 @@ tasks::is_abortable backup_task_impl::is_abortable() const noexcept {
     return tasks::is_abortable::yes;
 }
 
-void backup_task_impl::do_backup() {
-    if (!file_exists(_snapshot_dir.native()).get()) {
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
         throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
+    std::exception_ptr ex;
     gate uploads;
-    auto wait = defer([&uploads] { uploads.close().get(); });
-
     auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
-    auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
 
-    while (auto component_ent = snapshot_dir_lister.get().get()) {
```

Comment:
I would replace the `for (;;)` and the error handling of the `try/catch` inside it with:
```c++
try {
    while (auto component_ent = co_await snapshot_dir_lister.get()) {
    /* ...*/
    }
catch /* handle errors*/
```

Reply:
i explain the rationales:

> anyway, this would probably need to increase the indent level. so i am reverting to the coroutinized approach.

it's a compromise.